File: lesson9_general_problem_solving_approach/luhn.js
/*

input: number string
output: true/false based on if input is valid
rules:
  - input can contain spaces
  - counting from rightmost digit (check digit), and moving left, double every 2nd digit
    - for any digits that become >= 10, subtract 9 from result
    - ex:
      - 1111 => 2121
      - 8763 => 8 * 2, 7, 6 * 2, 3 => 16 - 9, 7, 12 - 9, 3 => 7, 7, 3, 3 => 7733
  - add all digits together to get checksum
    - ex:
      - 2121 => 2 + 1 + 2 + 1 => 6
      - 7733 => 20
  - if checksum % 10 === 0,  valid => true (ex: 6)
  - else, false (ex: 20)

- algo:
  - remove non digits from input
  - split input into array of digits
  - map digits arr to integers
  - reverse digits arr
  - map digits arr => (odd index digits * 2) - 9
  - reduce => sum digits
  - return sum % 10 === 0
*/


// const validLuhn = (numStr) => {
//   let digits = [...numStr.replace(/[^\d]/g, '')].map(Number).reverse();
//   digits = digits.map((d, i) => i % 2 === 1 ? d * 2 % 9: d);
//   let checksum = digits.reduce((sum, d) => sum + d, 0);
//   return checksum % 10 === 0;
// };


// const validLuhn = (numStr) => {
//   return !([...numStr.replace(/[^\d]/g, '')]
//     .reverse()
//     .map(Number)
//     .map((d, i) => i % 2 ? d * 2 % 9 : d)
//     .reduce((sum, d) => sum + d, 0) % 10);
// };


const validLuhn = (numStr) => {
  return !([...numStr.replace(/[^\d]/g, '')]
    .reverse().reduce((sum, d, i) => sum + (i % 2 ? d * 2 % 9 : +d), 0) % 10);
};


console.log(validLuhn('1111'));   // false
console.log(validLuhn('8763'));   // true
console.log(validLuhn('0'));      // true
console.log(validLuhn('10'));     // false
console.log(validLuhn('42'));     // 82 => 10 => true
console.log(validLuhn('123'));    // 143 => 8 => false
console.log(validLuhn('125'));    // 145 => 10 => true
console.log(validLuhn("2323 2005 7766 3554"));    // true
console.log(validLuhn("2323 2005 7766 3555"));    // false

console.log('=======');

console.log(validLuhn('1111') === false);
console.log(validLuhn('8763') === true);
console.log(validLuhn('0') ===    true);
console.log(validLuhn('10') ===   false);
console.log(validLuhn('42') ===   true);
console.log(validLuhn('123') ===  false);
console.log(validLuhn('125') ===  true);
console.log(validLuhn("2323 2005 7766 3554") ===  true);
console.log(validLuhn("2323 2005 7766 3555") ===  false);

File: launch-school-130/challenges/luhn.rb
# input: Integer
# output:
  # valid? method: return true/false
  # checksum method: return checksum
  # addends method:
    # append a check digit to make number valid, and return num + digit
      # ex: "2323 2005 7766 355" -> "2323 2005 7766 3554"
# rules:
  # rightmost digit = the check digit
  # valid:
    # counting from rightmost digit, double every 2nd digit
    # for any doubled digits that become > 10, subtract 9 from result
    # add all resulting digits together to get checksum
    # valid: checksum % 10 == 0
      # 1111 -> 2121 -> 6  -> invalid
      # 8763 -> 7733 -> 20 -> valid

class Luhn
  def self.create(num)
    "#{num}#{(0..9).find { |n| new("#{num}#{n}".to_i).valid? }}".to_i
  end

  def initialize(num)
    @digits = num.digits
  end

  def addends
    @digits.map.with_index { |n, i| i.even? ? n : ((n * 2) % 9.5).ceil }.reverse
  end

  def checksum
    addends.sum
  end

  def valid?
    (checksum % 10).zero?
  end
end

# p Luhn.new(12_121).addends
# p Luhn.new(8631).addends

# p Luhn.create(123)          # 1230
# p Luhn.create(873_956)      # 8_739_567
# p Luhn.create(837_263_756)  # 8_372_637_564

